concurrency/Chatbot.java

```diff
@@ -0,0 +1,1067 @@
+ * </ol>
+ *
+ * <b>Future ideas</b>
+ * <ul>
+ * <li>Configure the chat bot with JIDs of agents or admin that are allowed to:<ol>
+ *      <li>obtain workgroup and queue statistics</li>
+ *      <li>obtain current workgroup configureation</li>
+ *      <li>change the chatbot configuration</li>
+ *      <li>change the workgroup configuration, etc. etc. etc.</li></ol></li>
+ * <li>After the user ends the chat session with the agent offer the user to fill out
+ * a QoS survey.</li>
+ * </ul>
+ *
+ * @author Gaston Dombiak
+ */
+public class Chatbot implements UserCommunicationMethod {
+
+    private static final Logger Log = LoggerFactory.getLogger(Chatbot.class);
+    
+    /**
+     * Holds the workgroup where the chatbot is working. This is a one-to-one relation so this
+     * chatbot is the only chatbot that will be answering Messages sent to the workgroup.
+     */
+    private Workgroup workgroup;
+
+    /**
+     * The chat settings stores all the messages that the chatbot will use. The messages can
+     * be modified from the Admin Console.
+     */
+    private ChatSettings settings;
+
+    /**
+     * Holds the chat sessions of all the users that are trying to join this workgroup.
+     */
+    private Map<String, ChatbotSession> sessions = new ConcurrentHashMap<String, ChatbotSession>();
+
+    // TODO use resource bundles for these joining answers
+    /**
+     * Text that assumes that a user sent a positive answer.
+     */
+    private String yes = "yes";
+
+    /**
+     * Text that assumes that a user sent a negative answer.
+     */
+    private String no = "no";
+
+    /**
+     * Creates a new chatbot responsible for answering Messages sent to the specified workgroup.
+     *
+     * @param workgroup Workgroup where the new chatbot will be working.
+     */
+    public Chatbot(Workgroup workgroup) {
+        this.workgroup = workgroup;
+        this.settings = ChatSettingsManager.getInstance().getChatSettings(workgroup);
+    }
+
+    /**
+     * Returns the session of the specified user in a given workgroup. If no session exists then
+     * a new one will be created for the user if requested. If the session of the user remains
+     * inactive for some time then it may be discarded.
+     *
+     * @param user the user to return his session in a given workgroup.
+     * @param create true if a new session should be created if it doesn't already exist.
+     * @return the session of the specified user in a given workgroup.
+     */
+    public ChatbotSession getSession(JID user, boolean create) {
+        String fullJID = user.toString();
+        ChatbotSession session = sessions.get(fullJID);
+        if (session == null && create) {
+            synchronized (fullJID.intern()) {
+                session = sessions.get(fullJID);
+                if (session == null) {
+                    session = new ChatbotSession(user, this);
+                    sessions.put(fullJID, session);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 These lines use `ConcurrentHashMap`&#39;s **sequential** API and are protected by `synchronized` to ensure they execute together: **169 and 172**. This synchronization may reduce performance by limiting parallelism and does not take advantage of `ConcurrentHashMap`&#39;s **concurrent** API.

**Fix**  
 Consider replacing the check and `put()` with `putIfAbsent()` and then removing the synchronization. The check is at line: **170**.

**More info**  
 `putIfAbsent()` is thread-safe and puts the value only if the `ConcurrentHashMap` does not contain the key. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/HistoryFileManager.java#L927-L932) (external link).